### DISCUSSION TITLE
***2john on flatpak installation

### DISCUSSION THREAD

**Discussion by tufekk** [openwall/john-packages#151] [Q&A] [answered]
Hi,
Thank you for this awesome tool!
Forgive me if I missed the procedure somewhere but I wonder how to use, for example zip2john (or <anything-else>2john ) using flatpak installation.
There is a big bunch of /var/lib/flatpak/app/com.openwall.John/x86_64/stable/<very-long-numbers>/files/bin/***2john but I don't know how to call these options and don't find anything in the doc/help.

Have a good one

System: Debian testing
Flatpak: 1.14.4
Version: John the Ripper 1.9.0-jumbo-1 OMP [linux-gnu 64-bit x86_64 SSE2 AC]

**Comment by ghost** [comment, marked as answer]
Thanks for posting

Please, read about `zip2john` in the end of the post.

Unfortunately, there is no simple answer to your message because `*2john` tools have dependencies:
- at the moment some work better in Python 2;
- others have been ported and behave satisfactorily in Python 3;
- there are also `*2john` tools written in other languages;
- these tools have dependencies that also need to be installed for them to work correctly (Python modules, libraries, Perl modules, ...);
- these dependencies and version details are not written anywhere (that I know of).
 
The packages in this repository contain `john` ONLY (plus essential/required runtime libraries).
- I don't install Python;
- I don't install Lua or Rubu or ...
- I don't install extra Modules;

Why not?
- save space in these “packages” and save users’ internet bandwidth;
- I am NOT the right person to decide whether a `*2john` should use Python 2 or 3;
  - IMO, the `*2john` tools are a project themselves;
- I really like the idea of separating the processing of sensitive data files (via `*2john`) and the john session, which only needs a "hash".

Related to #19
***

Since `zip2john` is a symlink, it is possible to use it. The syntax is:
```
$ flatpak run --command=zip2john com.openwall.John <options>
```

Example:
```
#   wget https://openwall.info/wiki/_media/john/zip_sample_files.tar
#   mkdir -p examples/
#   tar -xvf zip_sample_files.tar --directory examples/

$ flatpak run --command=zip2john com.openwall.John examples/*.zip > examples.hash
```
Then:
```
$ flatpak run com.openwall.John examples.hash --pot=a1.pot --max-run-time=60 --format=ZIP
$ flatpak run com.openwall.John examples.hash --pot=a1.pot --max-run-time=60 --format=pkzip
```

Please, update your installation to `rolling-2310`. It is already available in FlatHub (soon, the bundle will be in this repo) and it is quite important for zip format.

**Reply by tufekk** [reply]
Thank you very much for your answer @claudioandre-br it worked!

Howerver I am a bit confused about the `rolling-2310` version you are talking about. How am I supposed to install that, is there a commit version I can specify to flatpak?

Indeed, after a "legacy" update with `flatpak update com.openwall.john` the app provide no output at all (commit: 481899b3317624a1a9df5ff202a8ec38b7748baa0bc9d0e4156203b8771ce824 on flathub repo).

Conversely, rolling back to the last commit f71957763afd5920c47f8b7b0c1a6fbd41ab1058c2dfd3bc65a4b8a9b02f0425 with a `flatpak update --comit=f71957763afd5920c47f8b7b0c1a6fbd41ab1058c2dfd3bc65a4b8a9b02f0425 com.openwall.John` provide me a functional app.

Maybe I should open another issue to this problem.

**Reply by ghost** [reply]
This is another issue.
- The root cause of the failure is that your CPU is very old. But, you will be able to use the new bundle successfully.

Unless you are a very lucky person, you need to update (using a package or building yourself) to use the zip format successfully.

---
To install a bundle, download the bundle file from the (not released yet) release here in GitHub, then
```
 # Your machine should have all the prerequisites installed.
#  sudo dnf install -y flatpak # or 'yum install', 'apt-get install', etc.
#  sudo flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo # flatpak repository
#  sudo flatpak install -y flathub org.freedesktop.Platform//23.08 # install the runtime (base "container")
 flatpak --user install --bundle john.flatpak # per-user installation (not system wide)
```

Anyway, you can open a new issue or discussion to talk about it.

**Reply by tufekk** [reply]
Very kind from you, I really appreciate the detail of your answers.

**Comment by solardiz** [comment]
> I am NOT the right person to decide whether a `*2john` should use Python 2 or 3

In cases where there's strict requirement for Python 2 or 3, we use explicit `python2` or `python3` in the shebang line. In cases when it's simply `python`, either is supposed to work, but how many/which Python modules need to also be installed for the script to work will vary by Python version.

So in most cases this has either already been decided or is not a critical decision (can be either), but some exceptions do exist. Overall, if the package provided some Python version or two consistent with the naming convention above, I think that would be an improvement.

**Comment by ghost** [comment]
BTW: 
* see also https://github.com/openwall/john/issues/5380
* the original post's CPU is SSE2 only, so the newly discovered bug doesn't matter.
